hail/src/main/scala/is/hail/shuffler/ShuffleClient.scala

```diff
@@ -0,0 +1,95 @@
+package is.hail.shuffler
+
+import org.apache.log4j.Logger;
+import is.hail.shuffler.ShuffleUtils._
+import is.hail.utils._
+import is.hail.annotations._
+import is.hail.expr.types.physical.{ PStruct, PType }
+import is.hail.expr.types.virtual.TStruct
+import is.hail.io.TypedCodecSpec
+import java.io._
+import java.net._
+import java.security.KeyStore;
+import java.util.UUID
+import java.util.concurrent.{ ConcurrentSkipListMap, Executors }
+import javax.net._
+import javax.net.ssl._
+import javax.security.cert.X509Certificate;
+import org.json4s.jackson.{ JsonMethods, Serialization }
+
+class ShuffleClient (
+  t: TStruct,
+  codecSpec: TypedCodecSpec,
+  key: Array[String],
+  ssl: SSLContext,
+  host: String,
+  port: Int
+) {
+  val log = Logger.getLogger(this.getClass.getName());
+
+  val sf = ssl.getSocketFactory()
+  val s = sf.createSocket(host, port)
+  val in = new DataInputStream(s.getInputStream())
+  val out = new DataOutputStream(s.getOutputStream())
+  log.info(s"CLNT connected to ${host}:${port}")
+
+  val keyedCodecSpec = new KeyedCodecSpec(t, codecSpec, key)
+  import keyedCodecSpec._
+
+  var uuid: String = null
+
+  def start(): Unit = {
+    log.info(s"CLNT start")
+    out.write(0)
+    Wire.writeTStruct(out, t)
+    Wire.writeTypedCodecSpec(out, codecSpec)
+    Wire.writeListOfStrings(out, key)
+    out.flush()
+    uuid = in.readUTF()
+    log.info(s"CLNT start done")
+  }
+
+  def put(values: Iterator[Long]): Unit = {
+    log.info(s"CLNT put")
+    val encoder = makeEnc(out)
+    encoder.writeByte(1)
+    encoder.flush()
+    out.writeUTF(uuid)
+    out.flush()
+    while (values.hasNext) {
+      encoder.writeByte(1)
+      val off = values.next
+      encoder.writeRegionValue(null, off)
+    }
+    encoder.writeByte(0)
+    encoder.flush()
+    assert(in.read() == 0)
+    log.info(s"CLNT put done")
+  }
+
+  def get(region: Region, l: Long, r: Long): Array[Long] = {
```

Comment:
Hmm. They are indeed ordered by key because of the LSM tree; however that's not obvious from the name of the tree. I've been iteratively refining a proposal on the dev forum which will eventually have a description of each operator. Those descriptions will definitely have information like "the results are ordered by key according to ordering O". Is that sufficient? I could also add a package-level README that described the design of the shuffler.